LoRaEngine/modules/LoRaWanNetworkSrvModule/LoRaWan.NetworkServer/ConcentratorDeduplication.cs

```diff
@@ -88,46 +85,37 @@ private bool EnsureFirstMessageInCache(string key, LoRaRequest loRaRequest, out
         }
 
-        internal static string CreateCacheKey(LoRaPayloadData payload)
+        internal static string CreateCacheKey(LoRaPayloadData payload, LoRaDevice loRaDevice)
         {
-            var totalBufferLength = DevAddr.Size + Mic.Size + (payload.RawMessage?.Length ?? 0) + payload.Fcnt.Length;
-            var buffer = totalBufferLength <= 128 ? stackalloc byte[totalBufferLength] : new byte[totalBufferLength]; // uses the stack for small allocations, otherwise the heap
-
-            var index = 0;
-            _ = payload.DevAddr.Write(buffer);
-            index += DevAddr.Size;
-
-            _ = payload.Mic is { } someMic ? someMic.Write(buffer[index..]) : throw new InvalidOperationException("Mic must not be null.");
-            index += Mic.Size;
-
-            payload.RawMessage?.CopyTo(buffer[index..]);
-            index += payload.RawMessage?.Length ?? 0;
-
-            BinaryPrimitives.WriteUInt16LittleEndian(buffer[index..], BinaryPrimitives.ReadUInt16LittleEndian(payload.Fcnt.Span));
+            var totalBufferLength = DevEui.Size + Mic.Size + payload.Fcnt.Length;
+            Span<byte> buffer = stackalloc byte[totalBufferLength];
+            var head = buffer; // keeps a view pointing at the start of the buffer
 
-            var key = Sha256.ComputeHash(buffer.ToArray());
+            buffer = DevEui.Parse(loRaDevice.DevEUI.AsSpan()).Write(buffer);
+            buffer = payload.Mic is { } someMic ? someMic.Write(buffer) : throw new InvalidOperationException("Mic must not be null.");
+            BinaryPrimitives.WriteUInt16LittleEndian(buffer, BinaryPrimitives.ReadUInt16LittleEndian(payload.Fcnt.Span));
 
-            return BitConverter.ToString(key);
+            return CreateCacheKeyCore(DataMessageCacheKeyPrefix, head);
         }
 
         internal static string CreateCacheKey(LoRaPayloadJoinRequest payload)
         {
-            var joinEui = JoinEui.Read(payload.AppEUI.Span);
-            var devEui = DevEui.Read(payload.DevEUI.Span);
-
             var totalBufferLength = JoinEui.Size + DevEui.Size + DevNonce.Size;
             Span<byte> buffer = stackalloc byte[totalBufferLength];
             var head = buffer; // keeps a view pointing at the start of the buffer
 
-            buffer = joinEui.Write(buffer);
-            buffer = devEui.Write(buffer);
+            buffer = JoinEui.Read(payload.AppEUI.Span).Write(buffer);
+            buffer = DevEui.Read(payload.DevEUI.Span).Write(buffer);
+            _ = payload.DevNonce.Write(buffer);
 
-            var devNonce = payload.DevNonce;
-            _ = devNonce.Write(buffer);
-
-            var key = Sha256.ComputeHash(head.ToArray());
+            return CreateCacheKeyCore(JoinMessageCacheKeyPrefix, head);
+        }
 
-            return BitConverter.ToString(key);
+        private static string CreateCacheKeyCore(string prefix, ReadOnlySpan<byte> buffer)
+        {
+            Span<char> key = stackalloc char[(buffer.Length * 3) - 1];
```

Comment:
I think you are making an assumption here about the length. Since we are using stackalloc it would be good to validate this to ensure we do not start using a larger buffer.

Reply:
Agree, changed in f1cb1a1e7011f4e140151315f8a601d7a873c295